### PR TITLE
Staking for direct staked nodes - part 2

### PR DESCRIPTION
**Pull request by ssd04**
## Reasoning behind the pull request
Added staking v4 integration testing scenarios:
- Withdraw unstaked funds before unbonding period should return error
- Withdraw unstaked funds in first available withdraw epoch
- Unstaking funds in different batches allows correct withdrawal for each batch at the corresponding epoch
- Unstake funds in different batches in the same epoch allows correct withdrawal in the correct epoch

## Testing procedure
- N/A

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: integrationTests/chainSimulator/staking/stakeAndUnStake_test.go [hunk 1]
```diff
@@ -1147,3 +1147,1143 @@ func testChainSimulatorDirectStakedUnstakeFundsWithDeactivationAndReactivation(t
...
+			NumNodesWaitingListMeta:  3,
+			NumNodesWaitingListShard: 3,
+			AlterConfigsFunction: func(cfg *config.Configs) {
+				cfg.EpochConfig.EnableEpochs.StakingV4Step1EnableEpoch = 100
+				cfg.EpochConfig.EnableEpochs.StakingV4Step2EnableEpoch = 101
+				cfg.EpochConfig.EnableEpochs.StakingV4Step3EnableEpoch = 102
+
+				cfg.EpochConfig.EnableEpochs.MaxNodesChangeEnableEpoch[2].EpochEnable = 102
+
+				cfg.SystemSCConfig.StakingSystemSCConfig.UnBondPeriodInEpochs = 6
```
Comment (iulianpascalau) [inline integrationTests/chainSimulator/staking/stakeAndUnStake_test.go:1688]: :+1:

File: integrationTests/chainSimulator/staking/stakeAndUnStake_test.go [hunk 2]
```diff
@@ -1147,3 +1147,1143 @@ func testChainSimulatorDirectStakedUnstakeFundsWithDeactivationAndReactivation(t
...
+	require.Nil(t, err)
+
+	testBLSKeyStaked(t, cs, metachainNode, blsKeys[0], targetEpoch)
+
+	shardIDValidatorOwner := cs.GetNodeHandler(0).GetShardCoordinator().ComputeId(validatorOwner.Bytes)
+	accountValidatorOwner, _, err := cs.GetNodeHandler(shardIDValidatorOwner).GetFacadeHandler().GetAccount(validatorOwner.Bech32, coreAPI.AccountQueryOptions{})
+	require.Nil(t, err)
+	balanceBeforeUnbonding, _ := big.NewInt(0).SetString(accountValidatorOwner.Balance, 10)
+
+	log.Info("Step 1. Create 3 transactions for unstaking: first one unstaking 11 egld each, second one unstaking 12 egld and third one unstaking 13 egld.")
```
Comment (iulianpascalau) [inline integrationTests/chainSimulator/staking/stakeAndUnStake_test.go:1832]: remove "each" ?

File: integrationTests/chainSimulator/staking/stakeAndUnStake_test.go [hunk 3]
```diff
@@ -1147,3 +1147,1143 @@ func testChainSimulatorDirectStakedUnstakeFundsWithDeactivationAndReactivation(t
...
+	unStakeValue1 = unStakeValue1.Mul(oneEGLD, unStakeValue1)
+	txDataField = fmt.Sprintf("unStakeTokens@%s", hex.EncodeToString(unStakeValue1.Bytes()))
+	txUnStake := generateTransaction(validatorOwner.Bytes, 1, vm.ValidatorSCAddress, zeroValue, txDataField, gasLimitForStakeOperation)
+	unStakeTx, err := cs.SendTxAndGenerateBlockTilTxIsExecuted(txUnStake, maxNumOfBlockToGenerateWhenExecutingTx)
+	require.Nil(t, err)
+	require.NotNil(t, unStakeTx)
+
+	unStakeTxFee, _ := big.NewInt(0).SetString(unStakeTx.Fee, 10)
+
+	epochIncr := int32(1)
```
Comment (iulianpascalau) [inline integrationTests/chainSimulator/staking/stakeAndUnStake_test.go:1845]: for easier read:
could have renamed this variable to `testEpoch`, initialized it to targetEpoch + 1 and then incremented this new variable wherever you increment epochIncr

File: integrationTests/chainSimulator/staking/stakeAndUnStake_test.go [hunk 4]
```diff
@@ -1147,3 +1147,1143 @@ func testChainSimulatorDirectStakedUnstakeFundsWithDeactivationAndReactivation(t
...
+	require.Nil(t, err)
+	require.Equal(t, okReturnCode, result.ReturnCode)
+
+	expectedStaked := big.NewInt(2600 - 11 - 12 - 13)
+	expectedStaked = expectedStaked.Mul(oneEGLD, expectedStaked)
+	require.Equal(t, expectedStaked.String(), string(result.ReturnData[0]))
+
+	log.Info("Step 3. Wait for the unbonding epoch to start")
+
+	epochIncr := int32(3)
```
Comment (iulianpascalau) [inline integrationTests/chainSimulator/staking/stakeAndUnStake_test.go:2252]: same here: testEpoch would read better

**Review by iulianpascalau — APPROVED** [review]

File: integrationTests/chainSimulator/staking/stakeAndUnStake_test.go (hunk 2 above)
Comment (ssd04) [reply integrationTests/chainSimulator/staking/stakeAndUnStake_test.go]: :+1:

File: integrationTests/chainSimulator/staking/stakeAndUnStake_test.go (hunk 3 above)
Comment (ssd04) [reply integrationTests/chainSimulator/staking/stakeAndUnStake_test.go]: refactored to use testEpoch variable

**Review by iulianpascalau — APPROVED** [review]

**Review by BeniaminDrasovean — APPROVED** [review]

**Pull request merged**